### PR TITLE
DBAAS-5538: Add Pipes as a first class feature store object

### PR DESCRIPTION
**Pull request by myles-novick**
## Description
New methods for getting, creating, updating, altering, and deleting pipes

## Motivation and Context
We need to support Pipes as a first class object in the feature store before being able to create pipelines and schedules in airflow. 
Fixes [DBAAS-5538](https://splicemachine.atlassian.net/browse/DBAAS-5538)

## Dependencies
[ml-workflow](https://github.com/splicemachine/ml-workflow/pull/193)

## How Has This Been Tested?
Tested against standalone db

### Additions
- New Pipe object
- New functions for getting, creating, updating, altering and deleting pipes

### Fixes
- Minor training view bugs

### CODE REVIEW

File: splicemachine/features/feature_store.py [hunk 1]
```diff
@@ -1043,6 +1049,99 @@ def remove_feature_set(self, schema_name: str, table_name: str, version: Union[s
...
+        assert name != "None", "Name of pipe cannot be None!"
+
+        f = base64.encodebytes(cloudpickle.dumps(function)).decode('ascii').strip()
+        p_dict = { "description": description, "function": f, "code": getsource(function) }
+
+        print(f'Updating Pipe {name} in the Feature Store')
+        r = make_request(self._FS_URL, f'{Endpoints.PIPES}/{name}', RequestType.PUT, self._auth, body=p_dict)
+        return Pipe(**r, splice_ctx=self.splice_ctx)
+
+    def alter_pipe(self, name: str, function: Optional[Union[str, Callable]] = None, description: Optional[str] = None) -> Pipe:
```
Comment (Ben-Epstein) [inline splicemachine/features/feature_store.py]: no version param

File: splicemachine/features/pipe.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    def apply(self, *args, **kwargs):   
+        func = cloudpickle.loads(self.function)
+        if self.language == 'pyspark':
+            func.__globals__['splice'] = self.splice_ctx 
+        r = func(*args, **kwargs)
+        if self.language == 'sql':
+            if not self.splice_ctx:
+                raise SpliceMachineException(f'Cannot execute SQL functions without a splice context')
+            return self.splice_ctx.df(r)
+        else: # 'python'
```
Comment (Ben-Epstein) [inline splicemachine/features/pipe.py:49]: if you're going to make this assumption, you should validate that the language passed in is `in ('python','sql','pyspark')`

File: splicemachine/features/feature_store.py [hunk 3]
```diff
@@ -1043,6 +1049,99 @@ def remove_feature_set(self, schema_name: str, table_name: str, version: Union[s
                      RequestType.DELETE, self._auth, { "schema": schema_name, "table":table_name, "version": version, "purge": purge })
         print('Done.')
 
+    def get_pipes(self, names: Optional[List[str]] = None) -> List[Pipe]:
```
Comment (Ben-Epstein) [inline splicemachine/features/feature_store.py:1051]: should we have a `get_pipe` function that takes a pipe (and optional version) ?

**Review by Ben-Epstein — CHANGES_REQUESTED** [review]
Few notes

File: splicemachine/features/pipe.py (hunk 2 above)
Comment (myles-novick) [reply splicemachine/features/pipe.py:49]: there's validation on the feature store side, but I can add it to client side too if you think its necessary

Comment (Ben-Epstein) [reply splicemachine/features/pipe.py:49]: I do. The scenario i'm thinking about is a user creates a Pipe here, and then tries to apply it before interacting with the server side. I would add it to the `__init__` in the language

**Review by Ben-Epstein — APPROVED** [review]

**Pull request merged**